model/flow/timeout_certificate.go

```diff
@@ -21,6 +27,63 @@ type TimeoutCertificate struct {
 	SigData crypto.Signature
 }
 
+// UntrustedTimeoutCertificate is an untrusted input-only representation of a TimeoutCertificate,
+// used for construction.
+//
+// This type exists to ensure that constructor functions are invoked explicitly
+// with named fields, which improves clarity and reduces the risk of incorrect field
+// ordering during construction.
+//
+// An instance of UntrustedTimeoutCertificate should be validated and converted into
+// a trusted TimeoutCertificate using NewTimeoutCertificate constructor.
+type UntrustedTimeoutCertificate TimeoutCertificate
+
+// NewTimeoutCertificate creates a new instance of TimeoutCertificate.
+// Construction TimeoutCertificate allowed only within the constructor.
+//
+// All errors indicate a valid TimeoutCertificate cannot be constructed from the input.
+func NewTimeoutCertificate(untrusted UntrustedTimeoutCertificate) (*TimeoutCertificate, error) {
+	if untrusted.NewestQC == nil {
+		return nil, fmt.Errorf("newest QC must not be nil")
+	}
+	if untrusted.SignerIndices == nil {
+		return nil, fmt.Errorf("signer indices must not be nil")
+	}
+	if untrusted.SigData == nil {
+		return nil, fmt.Errorf("signature must not be nil")
+	}
+
+	// The TC's view cannot be smaller than the view of the QC it contains.
+	// Note: we specifically allow for the TC to have the same view as the highest QC.
+	// This is useful as a fallback, because it allows replicas other than the designated
+	// leader to also collect votes and generate a QC.
+	if untrusted.View < untrusted.NewestQC.View {
+		return nil, fmt.Errorf("TC's QC view (%d) cannot be newer than the TC's view (%d)", untrusted.NewestQC.View, untrusted.View)
+	}
+
+	// verifying that tc.NewestQC is the QC with the highest view.
+	// Note: A byzantine TC could include `nil` for tc.NewestQCViews
+	if untrusted.NewestQCViews != nil {
```

Comment:
```suggestion
	if len(untrusted.NewestQCViews) > 0 {
```

Additionally, a TC with empty `NewestQCViews` is considered invalid. Moreover `len(NewestQCViews)` == `len(signers)` but we have no easy way to extract singers without parsing singer indices.